device/transport/mqtt/src/mqtt_twin_client.ts

```diff
@@ -0,0 +1,266 @@
+const responseTopic = '$iothub/twin/res/#';
+
+/* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_019: [** The subscribed topic for post events shall be $iothub/twin/PATCH/properties/desired/# **]** */
+const desiredPropertiesUpdatesTopic = '$iothub/twin/PATCH/properties/desired/#';
+
+interface TopicSubscription {
+  mqttClient: MqttBase;
+  topic: string;
+}
+
+/**
+ * @private
+ * @class        module:azure-iot-device-mqtt.MqttTwinReceiver
+ * @classdesc    Acts as a receiver for device-twin traffic
+ *
+ * @param {Object} config   configuration object
+ * @fires MqttTwinReceiver#subscribed   an MQTT topic has been successfully subscribed to
+ * @fires MqttTwinReceiver#error    an error has occured while subscribing to an MQTT topic
+ * @fires MqttTwinReceiver#response   a response message has been received from the service
+ * @fires MqttTwinReceiver#post a post message has been received from the service
+ * @throws {ReferenceError} If client parameter is falsy.
+ *
+ */
+export class MqttTwinClient extends EventEmitter {
+  static desiredPropertiesUpdateEvent: string = 'twinDesiredPropertiesUpdate';
+
+  private _mqtt: MqttBase;
+  private _pendingTwinRequests: { [key: string]: any } = {};
+  private _topicFsm: machina.BehavioralFsm;
+  private _responseTopic: TopicSubscription;
+  private _desiredPropertiesUpdatesTopic: TopicSubscription;
+
+  constructor(client: MqttBase) {
+    super();
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_18_001: [The `MqttTwinClient` constructor shall accept a `client` object.]*/
+    this._mqtt = client;
+
+    const messageHandler = this._onMqttMessage.bind(this);
+
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_001: [The `MqttTwinClient` constructor shall immediately subscribe to the `message` event of the `client` object.]*/
+    this._mqtt.on('message', messageHandler);
+
+    this._topicFsm = new machina.BehavioralFsm({
+      initialState: 'unsubscribed',
+      states: {
+        unsubscribed: {
+          _onEnter: function (client: TopicSubscription, err: Error, callback: (err?: Error) => void): void {
+            if (callback) {
+              callback(err);
+            }
+          },
+          subscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            this.transition(client, 'subscribing', callback);
+          },
+          unsubscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            // not entirely sure about that. if subscription are restored because cleanSession is false, it means technically a user may want to unsubscribe
+            // even though subscribe hasn't been called yet.
+            callback();
+          }
+        },
+        subscribing: {
+          _onEnter: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            client.mqttClient.subscribe(client.topic, { qos: 0 }, (err, result) => {
```

Comment:
>client.mqttClient.subscribe(client.topic [](start = 12, length = 40)

Ouch.  why alias this._mqtt to client.mqttClient? client.client.subscribe(client.topic) sounds like a 50's bubblegum song.  (client client bo-bient, banana fana fo-fient, fee fei fo-fient, client!)

As the code stands today, it would be clearer if you made _responseTopic and _desiredPropertiesUpdateTopic into strings and just skipped that little client object altogether

Reply:
Not sure that song ever made it overseas.

---
In reply to: [170700995](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170700995) [](ancestors = 170700995)

Reply:
... you have excellent code comments and poor musical taste. I hope the renaming helped a little bit.
Unfortunately we can't use strings in place of objects because the client/topicSubscription object is actually used to store the state of the behavioral state machine

---
In reply to: [170701724](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170701724) [](ancestors = 170701724,170700995)